src/Microsoft.AspNetCore.Routing/Matching/DfaMatcher.cs

```diff
@@ -226,11 +258,100 @@ public sealed override Task MatchAsync(HttpContext httpContext, EndpointFeature
+                LogLevel.Debug,
+                EventIds.CandidatesNotFound,
+                "No candidates found for the request path '{Path}'");
+
+            private static readonly Action<ILogger, int, string, Exception> _candidatesFound = LoggerMessage.Define<int, string>(
+                LogLevel.Debug,
+                EventIds.CandidatesFound,
+                "{CandidateCount} candidate(s) found for the request path '{Path}'");
+
+            private static readonly Action<ILogger, string, string, string, string, Exception> _candidateRejectedByComplexSegment = LoggerMessage.Define<string, string, string, string>(
+                LogLevel.Debug,
+                EventIds.CandidateRejectedByComplexSegment,
+                "Endpoint '{Endpoint}' with route pattern '{RoutePattern}' was rejected by complex segment '{Segment}' for the request path '{Path}'");
+
+            private static readonly Action<ILogger, string, string, string, object, string, Exception> _candidateRejectedByConstraint = LoggerMessage.Define<string, string, string, object, string>(
+                LogLevel.Debug,
+                EventIds.CandidateRejectedByConstraint,
+                "Endpoint '{Endpoint}' with route pattern '{RoutePattern}' was rejected by constraint '{ConstraintName}' with value '{RouteValue}' for the request path '{Path}'");
+
+            private static readonly Action<ILogger, string, string, string, Exception> _candidateNotValid = LoggerMessage.Define<string, string, string>(
+                LogLevel.Debug,
+                EventIds.CandidateNotValid,
+                "Endpoint '{Endpoint}' with route pattern '{RoutePattern}' is not valid for the request path '{Path}'");
+
+            private static readonly Action<ILogger, string, string, string, Exception> _candidateValid = LoggerMessage.Define<string, string, string>(
+                LogLevel.Debug,
+                EventIds.CandidateValid,
+                "Endpoint '{Endpoint}' with route pattern '{RoutePattern}' is valid for the request path '{Path}'");
+
+            public static void CandidatesNotFound(ILogger logger, string path)
+            {
+                _candidatesNotFound(logger, path, null);
+            }
+
+            public static void CandidatesFound(ILogger logger, string path, Candidate[] candidates)
+            {
+                _candidatesFound(logger, candidates.Length, path, null);
+            }
+
+            public static void CandidateRejectedByComplexSegment(ILogger logger, string path, Endpoint endpoint, RoutePatternPathSegment segment)
+            {
+                // This should return a real pattern since we're processing complex segments.... but just in case.
+                var routePattern = GetRoutePattern(endpoint);
+                _candidateRejectedByComplexSegment(logger, endpoint.DisplayName, routePattern, segment.DebuggerToString(), path, null);
```

Comment:
Usually a log level check happens inside the method when doing something allocaty. Add a comment that the DebuggerToString is ok because we've already checked the log level.

Reply:
Actually this method isn't wrapped in an `if (log)` check. Should add it here.

Reply:
Yeah, this should be 👍